*: Add initial remote-write-receive component

brancz opened the pull request:
<!--
    Keep PR title verbose enough and add prefix telling
    about what components it touches e.g "query:" or ".*:"
-->

## Changes

This adds the very first pieces of the [proposal to add receiving remote-write](https://github.com/improbable-eng/thanos/blob/master/docs/proposals/approved/201812_thanos-remote-receive.md) support to Thanos. The features are just having a single tsdb that remote write requests are written into and that the component joins a mesh network to enable querying. All other features are still outstanding but can be worked on in parallel once this lands.

## Verification

As this is very small and primarily hooks up existing components, the extend of my verification is that I added this new component to the quickstart script and verified that indeed data replicated to the endpoint receiving remote-write calls can be queried via the Thanos querier. Happy to add this to the e2e test suite though.

@bwplotka @domgreen 

cc @metalmatze @mxinden @squat @s-urbaniak @ant31

bwplotka commented:
Finally <3

First of all, as you can see our CI does not like NOT checked errors (: and both:
```
fmt.Fprintf	fmt.Fprintf(w, "Service Unavailable")
(*net/http.Server).Shutdown	httpSrv.Shutdown(ctx)
```
looks like it would be nice at least log them.

brancz commented:
Looks like that made CI happy :slightly_smiling_face: .

brancz commented:
I think having a simple e2e test added would be good. I'll look into adding one.

bwplotka commented on cmd/thanos/remote-write-receive.go:
```diff
@@ -0,0 +1,281 @@
+)
+
+func registerRemoteWriteReceive(m map[string]setupFunc, app *kingpin.Application, name string) {
+	cmd := app.Command(name, "accept Prometheus remote write API requests (EXPERIMENTAL, this may change drastically without notice)")
```
```suggestion
	cmd := app.Command(name, "Accept Prometheus remote write API requests (EXPERIMENTAL, this may change drastically without notice)")
```

bwplotka commented on cmd/thanos/remote-write-receive.go:
```diff
@@ -0,0 +1,281 @@
+				return nil
+			},
+			func(err error) {
+				level.Debug(logger).Log("msg", "initial load group errored", "err", err)
```
hmm, I don't get this log messages here and on every group. This code is invoked even though this group did not errof (but e.g other group errored). I think we can remove those debug logs. We will print error for problematic group via `g.Run` anyway.

brancz replied:
Sorry, this is a leftover from some debugging, I'll remove it.

bwplotka commented on cmd/thanos/remote-write-receive.go:
```diff
@@ -0,0 +1,281 @@
+			},
+			func(err error) {
+				level.Debug(logger).Log("msg", "tsdb group errored", "err", err)
+				if err := localStorage.Close(); err != nil {
```
just checking .. localStorage.Close closes `db` as well?

brancz replied:
it does

bwplotka commented on cmd/thanos/remote-write-receive.go:
```diff
@@ -0,0 +1,281 @@
+			func(err error) {
+				level.Debug(logger).Log("msg", "tsdb group errored", "err", err)
+				if err := localStorage.Close(); err != nil {
+					level.Error(logger).Log("msg", "Error stopping storage", "err", err)
```
all log lines should start lowercase

bwplotka commented on cmd/thanos/remote-write-receive.go:
```diff
@@ -0,0 +1,281 @@
+
+	level.Debug(logger).Log("msg", "setting up metric http listen-group")
+	if err := metricHTTPListenGroup(g, logger, reg, httpMetricsBindAddr); err != nil {
+		level.Debug(logger).Log("msg", "metric listener errored", "err", err)
```
Replicated error handling. (rule "Handle error just once") (: 

I think error wrap is the way to go (and removal of this log line).

bwplotka commented on cmd/thanos/remote-write-receive.go:
```diff
@@ -0,0 +1,281 @@
+		g.Add(
+			func() error {
+				level.Info(logger).Log("msg", "Starting TSDB ...")
+				db, err := tsdb.Open(
```
hm.. can we open DB outside of run group? And avoid blocking on `dbOpen` entirely? (:

brancz replied:
The point here is that we respond to http requests as early as possible, but with `StatusServiceUnavailable` until tsdb is loaded.

bwplotka commented on pkg/receive/handler.go:
```diff
@@ -0,0 +1,222 @@
+	ReadyStorage   *tsdb.ReadyStorage
+}
+
+// Handler serves various HTTP endpoints of the Prometheus server
```
Missing trailing period. Also `various` is kind of wide word..  I think we expose only `receive` right? (:

bwplotka commented on pkg/receive/handler.go:
```diff
@@ -0,0 +1,222 @@
+		ReadTimeout: h.options.ReadTimeout,
+	}
+
+	errCh := make(chan error)
```
I am pretty sure we leak go routine here, as we send twice but read only once.

Worth to use `run.Group` here for easier code?

metalmatze replied:
Yes, this should be a `run.Group`.

bwplotka replied:
still valid?

bwplotka replied:
Not addressed

bwplotka commented on pkg/receive/handler.go:
```diff
@@ -0,0 +1,222 @@
+
+	var (
+		m     = cmux.New(listener)
+		httpl = m.Match(cmux.HTTP1Fast())
```
Can you comment what this is for? It might be not straightforward.

bwplotka commented on pkg/receive/handler.go:
```diff
@@ -0,0 +1,222 @@
+		return
+	}
+
+	err = h.receiver.Receive(&wreq)
```
we can do `if err := ... ; err != nil` as well

bwplotka commented on pkg/receive/receivewriter.go:
```diff
@@ -0,0 +1,57 @@
+package receive
+
+import (
+	"github.com/go-kit/kit/log"
```
`make format` needed (:

bwplotka commented on scripts/prometheus.yml:
```diff
@@ -0,0 +1,10 @@
+# When the Thanos remote-write-receive component is started,
```
Hm.. can we add this to `remote-write-receive` command description instead?  Might be easier to discover

bwplotka commented on scripts/quickstart.sh:
```diff
@@ -136,4 +142,22 @@ do
 
+sleep 0.5
+
+if [ -n "${REMOTE_WRITE_ENABLED}" ]
```
Wow I did not know that this script still works (:

bwplotka commented:
And yea, some tests would be nice. (:

metalmatze commented on cmd/thanos/remote-write-receive.go:
```diff
@@ -0,0 +1,281 @@
+
+	var metadata = &promMetadata{
+		// Start out with the full time range. The shipper will constrain it later.
+		// TODO(fabxc): minimum timestamp is never adjusted if shipping is disabled.
```
Was this code copied from somewhere? Slightly confusing to see `TODO(fabxc)` in here, but not a big deal I guess.

bwplotka replied:
haha yea was surprised as well at  the beginning (: It is copied indeed.

metalmatze commented on pkg/receive/handler.go:
```diff
@@ -0,0 +1,222 @@
+// Options for the web Handler.
+type Options struct {
+	Receiver       *ReceiveWriter
+	Context        context.Context
```
This can be removed as it is not used.

metalmatze commented on pkg/receive/handler.go:
```diff
@@ -0,0 +1,222 @@
+// Handler serves various HTTP endpoints of the Prometheus server
+type Handler struct {
+	readyStorage *tsdb.ReadyStorage
+	context      context.Context
```
This can be removed as it is not used.

metalmatze commented on pkg/receive/handler.go:
```diff
@@ -0,0 +1,222 @@
+func (h *Handler) receive(w http.ResponseWriter, req *http.Request) {
+	defer func() {
+		if r := recover(); r != nil {
+			fmt.Println("panic recovered:", r, string(debug.Stack()))
```
Maybe we want to have a counter for how many panics were recovered?

metalmatze commented on pkg/receive/receivewriter.go:
```diff
@@ -0,0 +1,57 @@
+	Appender() (storage.Appender, error)
+}
+
+type ReceiveWriter struct {
```
This is spelled `receive.ReceiveWriter` and should be renamed to `receive.Writer` to be more idiomatic.

brancz commented:
@bwplotka @metalmatze I believe I addressed everything or answered the comments and added the receiver to the e2e test setup (the test is that a prometheus is successfully pushing an `up` metric of a "down" node-exporter).

domgreen commented on cmd/thanos/main.go:
```diff
@@ -78,6 +78,7 @@ func main() {
 	registerCompact(cmds, app, "compact")
 	registerBucket(cmds, app, "bucket")
 	registerDownsample(cmds, app, "downsample")
+	registerRemoteWriteReceive(cmds, app, "remote-write-receive")
```
nit - would shorten this to just `receive`/`remote` /`receiver`

domgreen commented on cmd/thanos/remote-write-receive.go:
```diff
@@ -0,0 +1,270 @@
+			cancel()
+			peer.Close(2 * time.Second)
+		})
+	}
```
can we remove this whole block? Will receive need to peer via gossip? If we are removing gossip in the near future I think we should not add it to any new components and instead help people migrate to the new approach.

domgreen commented on cmd/thanos/remote-write-receive.go:
```diff
@@ -0,0 +1,270 @@
+	level.Debug(logger).Log("msg", "setting up grpc server")
+	{
+		var (
+			logger = log.With(logger, "component", "receiver")
```
`receiver` > `receive`? or the verbose name? Let's keep it consistant.

domgreen commented on pkg/component/component.go:
```diff
@@ -71,17 +71,20 @@ func FromProto(storeType storepb.StoreType) StoreAPI {
+	Rule        = sourceStoreAPI{component: component{name: "rule"}}
+	Sidecar     = sourceStoreAPI{component: component{name: "sidecar"}}
+	Store       = sourceStoreAPI{component: component{name: "store"}}
+	RemoteWrite = sourceStoreAPI{component: component{name: "remote_write"}}
```
consistency

domgreen commented on pkg/receive/handler.go:
```diff
@@ -0,0 +1,197 @@
+
+	reqBuf, err := snappy.Decode(nil, compressed)
+	if err != nil {
+		fmt.Println("snappy decode error")
```
`fmt` > `logger`?

domgreen commented on scripts/quickstart.sh:
```diff
@@ -136,4 +142,35 @@ do
+
+if [ -n "${REMOTE_WRITE_ENABLED}" ]
+then
+  ./thanos remote-write-receive \
```
:+1: this is great taht were adding it here as well.

brancz replied:
either we leave this or remove gossip :slightly_smiling_face:

domgreen replied:
@bwplotka what do you think on this ... personally, I think we should phase out gossip so (in a different PR) change this file to work via the new approach. 
So for this PR I would remove `receive` from this and once the above work has been done add it back in.

brancz replied:
As we have the e2e tests, that works for me.

brancz replied:
As there is no git-history of this, I'm going to comment this out for now, but leave it. Does that sound like an ok compromise?

domgreen replied:
I'm happy with that ... can you make sure a comment about why its commented is included :smile:

brancz replied:
Sounds good to me

domgreen commented on cmd/thanos/remote-write-receive.go:
```diff
@@ -0,0 +1,270 @@
+	dataDir string,
+	peer cluster.Peer,
+) error {
+	level.Info(logger).Log("msg", "setting up receiver")
```
Is it worth adding a warn log here saying that it is currently experimental?

brancz replied:
very good idea :+1:

bwplotka commented on pkg/receive/handler.go:
```diff
@@ -0,0 +1,203 @@
+var (
+	requestDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "prometheus_http_request_duration_seconds",
```
We might want `thanos` prefix for those metrics? Or do we want to keep it similar to get advantage of common dashboards?

brancz replied:
Sorry that was a copy paste mistake. Yes lets make it `thanos_http_...`.

bwplotka commented on pkg/receive/handler.go:
```diff
@@ -0,0 +1,203 @@
+	return func(w http.ResponseWriter, r *http.Request) {
+		if h.isReady() {
+			f(w, r)
+		} else {
```
just return here, no need for `else`

bwplotka commented on cmd/thanos/receive.go:
```diff
@@ -0,0 +1,271 @@
+		Default("./data").String()
+
+	m[name] = func(g *run.Group, logger log.Logger, reg *prometheus.Registry, tracer opentracing.Tracer, _ bool) error {
+		peer, err := newPeerFn(logger, reg, false, "", false)
```
You can actualy remove gossip stuff if you want, up to you - making sure deprecated stuff is in new component is quite not necessary. (:  Not a blocker

bwplotka commented on cmd/thanos/receive.go:
```diff
@@ -0,0 +1,271 @@
+	dataDir string,
+	peer cluster.Peer,
+) error {
+	level.Info(logger).Log("msg", "setting up receive")
```
Let's kill this message, below as `setting up receive; this component is ...`

bwplotka commented on cmd/thanos/receive.go:
```diff
@@ -0,0 +1,271 @@
+				select {
+				case <-dbOpen:
+					break
+				// In case a shutdown is initiated before the dbOpen is released
```
Missing trailing period in comment (:

bwplotka replied:
Also I would remove this, I think this is quite clear.

bwplotka commented on cmd/thanos/receive.go:
```diff
@@ -0,0 +1,271 @@
+	level.Debug(logger).Log("msg", "setting up grpc server")
+	{
+		var (
+			logger = log.With(logger, "component", "receiver")
```
Why not doing this in very top?

bwplotka commented on go.sum:
```diff
@@ -204,6 +204,7 @@ github.com/onsi/ginkgo v1.6.0 h1:Ix8l273rp3QzYgXSR+c8d1fTG7UPgYkOSELPhiY/YGw=
 github.com/onsi/ginkgo v1.6.0/go.mod h1:lLunBs/Ym6LB5Z9jYTR76FiuTmxDTDusOGeTQH+WWjE=
 github.com/onsi/gomega v1.4.1 h1:PZSj/UFNaVp3KxrzHOcS7oyuWA7LoOY/77yCTEFu21U=
 github.com/onsi/gomega v1.4.1/go.mod h1:C1qb7wdrVGGVU+Z6iS04AVkA3Q65CEZX59MT0QO5uiA=
+github.com/opentracing-contrib/go-stdlib v0.0.0-20170113013457-1de4cc2120e7 h1:8KbikWulLUcMM96hBxjgoo6gTmCkG6HYSDohv/WygYU=
```
modules :heart:

bwplotka commented on pkg/receive/handler.go:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	readyf := h.testReady
+	router.Post("/receive", readyf(h.receive))
```
Ok, so this starts to be some API. We already keep Prometheus like API to guide people and be intuitive. This means that remote receiver would be nice to use the same or similar... does Prometheus serves remote write endpoint? What endpoint looks like?  `api/v1/write` ? As a basics something like `v1` in path is useful.

domgreen replied:
tend to agree but flexible on the url ... `api/v1/receive` feels more correct to me

brancz replied:
agreed with the prefix, going with `/api/v1/receive`

brancz commented:
I believe I addressed everything:

* Added run group handling for the http handler
* Removed gossip
* Took the freedom to convert the quickstart script from gossip to static `--store` flag configuration

bwplotka commented on cmd/thanos/flags.go:
```diff
@@ -17,25 +17,37 @@ import (
 )
 
-func regCommonServerFlags(cmd *kingpin.CmdClause) (
+func regGRPCFlags(cmd *kingpin.CmdClause) (
```
hahah I was starring at monitor 1m and thinking why you removed Gossip for everyone in this PR... then seen that actually you added this function.... Github troll =D
